app/src/main/java/ca/llamabagel/transpo/data/TripsRepository.kt

```diff
@@ -4,29 +4,90 @@
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
 import kotlinx.coroutines.withContext
 import java.io.IOException
 import javax.inject.Inject
+import javax.inject.Singleton
 
+@Singleton
+@ExperimentalCoroutinesApi
 class TripsRepository @Inject constructor(
     private val database: TransitDatabase,
-    private val apiService: TripsService
+    private val apiService: TripsService,
+    private val sharedPreferences: SharedPreferences
 ) {
+    private val cachedResults: MutableMap<StopCode, ConflatedBroadcastChannel<ApiResponse>> = mutableMapOf()
+    private val cachedStopCodes: MutableMap<StopId, StopCode> = mutableMapOf()
 
-    suspend fun getStop(stopId: String): Stop? = withContext(Dispatchers.IO) {
-        database.stopQueries.getStopById(stopId).executeAsOneOrNull()
+    suspend fun getResultCache(stopId: StopId): ConflatedBroadcastChannel<ApiResponse> = withContext(Dispatchers.IO) {
+        val stopCode =
+            cachedStopCodes.getOrPut(stopId, { database.stopQueries.getStopById(stopId).executeAsOne().code })
+
+        cachedResults.getOrPut(stopCode, { ConflatedBroadcastChannel() })
+    }
+
+    fun clearCacheFor(stopId: StopId) {
+        cachedStopCodes[stopId]?.let(cachedResults::remove)
+    }
+
+    suspend fun getStop(stopId: StopId): Result<Stop> = withContext(Dispatchers.IO) {
+        try {
+            val stop = database.stopQueries.getStopById(stopId).executeAsOne()
+
+            // Cache the stop code for other uses
+            if (!cachedStopCodes.containsKey(stopId)) {
+                cachedStopCodes[stopId] = stop.code
+            }
+
+            Result.Success(stop)
+        } catch (e: Exception) {
+            Result.Error(e)
+        }
     }
 
-    suspend fun getTrips(stopCode: String): Result<ApiResponse> = withContext(Dispatchers.IO) {
+    suspend fun getTrips(stopId: StopId): Result<Unit> = withContext(Dispatchers.IO) {
         try {
-            Result.Success(apiService.getTrips(stopCode).await())
+            val stopCode =
+                cachedStopCodes.getOrPut(stopId, { database.stopQueries.getStopById(stopId).executeAsOne().code })
+
+            cachedResults.getOrPut(stopCode, { ConflatedBroadcastChannel() })
+                .offer(apiService.getTrips(stopCode.value).await())
+            Result.Success(Unit)
         } catch (e: IOException) {
             Result.Error(e)
         }
     }
+
+    @SuppressLint("ApplySharedPref")
+    suspend fun setGroupByRoute(group: Boolean) = withContext(Dispatchers.IO) {
+        sharedPreferences.edit().putBoolean(KEY_GROUP_BY_ROUTE, group).commit()
+        rebroadcast()
+    }
+
+    suspend fun getGroupByRoute(): Boolean =
+        withContext(Dispatchers.IO) { sharedPreferences.getBoolean(KEY_GROUP_BY_ROUTE, false) }
+
+    /**
+     * Rebroadcasts the last response
+     */
+    private fun rebroadcast() {
+        cachedResults.forEach { (_, v) ->
```

Comment:
I think it's a bit more readable if we only have 1 parameter in the `forEach`. Also maybe change `v` to something more descriptive?
```suggestion
        cachedResults.map { it.value }.forEach { v -> v.offer(v.value) }
```

Reply:
`.map { it.value }` is just the verbose version of `(_, v)`. Not very Kotlin-like. Plus doing a map creates a wasteful temporary object. I can give it a better name though.

Reply:
I would just use `.forEachValue` but that requires Java 8 (i.e. SDK 24+)